x/hold/events_test.go

```diff
@@ -0,0 +1,159 @@
+package hold
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	abci "github.com/tendermint/tendermint/abci/types"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/provenance-io/provenance/testutil/assertions"
+)
+
+func TestNewEventHoldAdded(t *testing.T) {
+	tests := []struct {
+		name   string
+		addr   sdk.AccAddress
+		amount sdk.Coins
+		reason string
+		exp    *EventHoldAdded
+	}{
+		{
+			name:   "both nil",
+			addr:   nil,
+			amount: nil,
+			exp:    &EventHoldAdded{Address: "", Amount: ""},
+		},
+		{
+			name:   "both empty",
+			addr:   sdk.AccAddress{},
+			amount: sdk.Coins{},
+			exp:    &EventHoldAdded{Address: "", Amount: ""},
+		},
+		{
+			name:   "normal address and two denoms",
+			addr:   sdk.AccAddress("normal_address______"),
+			amount: sdk.NewCoins(sdk.NewInt64Coin("fingercoin", 10), sdk.NewInt64Coin("toecoin", 9)),
+			exp: &EventHoldAdded{
+				Address: sdk.AccAddress("normal_address______").String(),
+				Amount:  "10fingercoin,9toecoin",
+			},
+		},
+		{
+			name:   "only a reason",
+			reason: "this is a test reason",
+			exp:    &EventHoldAdded{Reason: "this is a test reason"},
+		},
+		{
+			name:   "control",
+			addr:   sdk.AccAddress("control_address_____"),
+			amount: sdk.NewCoins(sdk.NewInt64Coin("cherry", 4)),
+			reason: "control reason",
+			exp: &EventHoldAdded{
+				Address: sdk.AccAddress("control_address_____").String(),
+				Amount:  sdk.NewCoins(sdk.NewInt64Coin("cherry", 4)).String(),
+				Reason:  "control reason",
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			event := NewEventHoldAdded(tc.addr, tc.amount, tc.reason)
+			assert.Equal(t, tc.exp, event, "NewEventHoldAdded")
+		})
+	}
+}
+
+func TestNewEventHoldReleased(t *testing.T) {
+	tests := []struct {
+		name   string
+		addr   sdk.AccAddress
+		amount sdk.Coins
+		exp    *EventHoldReleased
+	}{
+		{
+			name:   "both nil",
+			addr:   nil,
+			amount: nil,
+			exp:    &EventHoldReleased{Address: "", Amount: ""},
+		},
+		{
+			name:   "both empty",
+			addr:   sdk.AccAddress{},
+			amount: sdk.Coins{},
+			exp:    &EventHoldReleased{Address: "", Amount: ""},
+		},
+		{
+			name:   "normal address and two denoms",
+			addr:   sdk.AccAddress("normal_address______"),
+			amount: sdk.NewCoins(sdk.NewInt64Coin("fingercoin", 10), sdk.NewInt64Coin("toecoin", 9)),
+			exp: &EventHoldReleased{
+				Address: sdk.AccAddress("normal_address______").String(),
+				Amount:  "10fingercoin,9toecoin",
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			event := NewEventHoldReleased(tc.addr, tc.amount)
+			assert.Equal(t, tc.exp, event, "NewEventHoldReleased")
+		})
+	}
+}
+
+func TestTypedEventToEvent(t *testing.T) {
```

Comment:
Why do we have to test the sdk.TypedEvenToEvent functionality?

Reply:
We aren't actually testing `sdk.TypedEvenToEvent`, we're testing our new events being provided to that function.

The only reason that `EmitTypedEvent` returns an error is if `TypedEvenToEvent` returns an error. `TypedEvenToEvent` will only return an error if there's json marshal/unmarshal problems which will only happen if there's something structurally wrong with the event type. Basically, if `TypedEvenToEvent` returns an error for one instance of a type, it'll return an error for all instances of that type.

Since these event types are designed to be provided to `EmitTypedEvent` (and thus `TypedEventToEvent`), I felt it prudent to make sure they can be safely converted.

I was even considering ignoring the error from `EmitTypedEvent` because of this, but couldn't quite bring myself to do it. I'd bet that the funcs that call `EmitTypedEvent` in here already need to return an error field anyway, so doing that probably doesn't buy me anything.